shared/js/intl_helper.js

```diff
@@ -0,0 +1,148 @@
+        objectCache = new Map();
+      }
+
+      if (objectCache.has(name)) {
+        throw new Error(
+          'Intl Object with name "' + name + '" already exists');
+      }
+
+      objectCache.set(name, {
+        options,
+        type,
+        object: undefined,
+        listeners: new Set()
+      });
+    },
+
+    getObject: function(name) {
+      if (!objectCache.has(name)) {
+        throw new Error('Intl Object with name "' + name + '" is not defined');
+      }
+      const obj = objectCache.get(name);
+
+      if (obj.object === undefined) {
+        switch (obj.type) {
+          case 'number':
+            obj.object = Intl.NumberFormat(
+              navigator.languages,
+              obj.options
+            );
+            break;
+          case 'datetime':
+            const options = Object.assign({}, obj.options);
+            if (options.hour) {
+              options.hour12 = navigator.mozHour12;
+            }
+            console.log('new Format');
+            console.log(options);
+            obj.object = Intl.DateTimeFormat(
+              navigator.languages,
+              options
+            );
+            break;
+          default:
+            throw new Error('Unknown type: ' + obj.type);
+        }
+      }
+      return obj.object;
+    },
+    on: function(names, cb) {
+      if (typeof names === 'string') {
+        names = [names];
+      }
+      for (var name of names) {
+        if (!objectCache.has(name)) {
+          throw new Error(
+            'Intl Object with name "' + name + '" is not defined');
+        }
+
+        var obj = objectCache.get(name);
+
+        obj.listeners.add(cb);
+      }
+    },
+    off: function(name, cb) {
+      var obj = objectCache.get(name);
+
+      obj.listeners.remove(cb);
+    },
+    emit: function(type, ...args) {
+      var affectedCallbacks = new Set();
+
+      objectCache.forEach((obj, name) => {
+        if (obj.type === type) {
+          obj.listeners.forEach(cb => affectedCallbacks.add(cb));
+        }
+      });
+
+      affectedCallbacks.forEach(cb => {
+        cb(...args);
+      });
+    }
+  };
+
+  function resetDateTime() {
+    // XXX: we could be smarter here and reset on timeformatchange only
+    // if the formatter uses 'hour' option
+
+    objectCache.forEach((obj, name) => {
+      if (obj.type === 'datetime') {
+        obj.object = undefined;
+      }
+    });
+  }
+
+  function resetNumber(reason) {
+    objectCache.forEach((obj, name) => {
+      if (obj.type === 'number') {
+        obj.object = undefined;
+      }
+    });
+  }
+
+  window.addEventListener('timeformatchange', () => {
+    resetDateTime();
+    global.IntlHelper.emit('datetime', {reason: 'timeformatchange'});
+  });
+
+  window.addEventListener('languagechange', () => {
+    resetDateTime();
+    resetNumber();
+    global.IntlHelper.emit('number', {reason: 'languagechange'});
+    global.IntlHelper.emit('datetime', {reason: 'languagechange'});
```

Comment:
right now yes, but soon I'll land bug 1187539 which makes us not fire timeformatchange on languagechange, so I'm counting on that.